xbmc/filesystem/PluginDirectory.cpp

```diff
@@ -544,6 +559,11 @@ bool CPluginDirectory::WaitOnScriptResult(const std::string &scriptPath, int scr
   return !m_cancelled && m_success;
 }
 
+bool CPluginDirectory::ProcessRenderLoop(bool renderOnly)
+{
+  return CServiceBroker::GetGUI()->GetWindowManager().ProcessRenderLoop(renderOnly);
```

Comment:
Do you really want filesystem calling into GUI? This way you'll never get headless and make it harder to split the render thread.
Since ProcessRenderLoop is a hack by itself, there might be more issues related. The most obvious one is that you slow down rendering of the first caller by creating a 2nd loop.

Reply:
> Do you really want filesystem calling into GUI?

 This is a nogo, imo.

Reply:
Actually this already happens by abusing CGUIDialogBusy. Instead of making it worse I would get rid of CGUIDialogBusy here.

Reply:
> Instead of making it worse I would get rid of CGUIDialogBusy here.

+1

Reply:
<pre>
#include "dialogs/GUIDialogBusy.h"
#include "dialogs/GUIDialogProgress.h"
#include "guilib/GUIComponent.h"
#include "guilib/GUIWindowManager.h"
</pre>

All GUI stuff which does not belong here...

Reply:
Ok guys - so what's the right approach to do it?
If I get the concept correctly, most of the `CPluginDirectory` operations are about running a python script in the background and waiting for it's result.
So the `BusyDialog` makes perfect sense and allows the user to cancel fetching of items for example.

>Actually this already happens by abusing CGUIDialogBusy. Instead of making it worse I would get rid of CGUIDialogBusy here.

Given above thought, if I just remove the BusyDialog the user won't have any control over these operations anymore.
That leads me to this question:
A filesystem or vfs operation is often a timeconsuming thing and something where we want to show progress/spinners or whatever in the GUI.
Do we already have infrastructure in code or a plan in mind on how to do that in the future?

Reply:
I guess it would also make sense to pickup: https://github.com/xbmc/xbmc/pull/15825 ?

Reply:
I'm not aware of a concept, sorry. IMO, busy spinners, progress bars and all other GUI stuff should be implemented by the code calling vfs functions, not vfs itself.

Reply:
Yeah exactly, it should actually be the other way round.
The code that triggers vfs functionality would possibly spwan it in the background and render something until vfs completes.

Reply:
For PVR GUI actions, I did it like that: https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/PVRGUIActions.cpp#L82 ff.

Reply:
I see, thx for the pointer.

Reply:
> The code that triggers vfs functionality would possibly spwan it in the background and render something until vfs completes.

that would be really bad. the caller should not be aware of implementation details of the service. just make the call asynchronous: trigger an action, go to a waiting state, receive event that action has finished, process the result.